Don't access `NULL` GC delta

chesedo opened the pull request:
## Describe the changes in the pull request
The GC delta was freed from memory when it was applied a few lines before the `resetCardinality` call. This changes gets the last block ID in the delta before this apply call which frees it, and then passes this last block to `resetCardinality` instead.

This is for this redisearch `master` build error log -> https://github.com/redis/redis/issues/14570#issuecomment-3581813475

#### Mark if applicable

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Prevents NULL access by computing numeric start position before applying GC deltas, corrects HLL reset to include ignored block, and standardizes delta ownership by nulling after apply and unconditionally freeing across handlers.
> 
> - **Numeric index/HLL**:
>   - Compute `blocksSinceFork` before `InvertedIndex_ApplyGcDelta` and pass to `resetCardinality`.
>   - In `resetCardinality`, include ignored block (`blocksSinceFork++`) and derive `startIdx` via `InvertedIndex_NumBlocks(...) - blocksSinceFork`.
> - **Delta ownership/lifecycle**:
>   - After `InvertedIndex_ApplyGcDelta(...)`, set `delta = NULL`/`ninfo->delta = NULL` and unconditionally `InvertedIndex_GcDelta_Free(delta)` in cleanup.
>   - Remove `shouldFreeDeltas` branches in term/tag/missing/existing docs handlers, standardizing freeing behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d42191025d8f0fdf59a7990638ac80bc307c34ba. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

GuyAv46 commented on src/fork_gc.c:
```diff
@@ -508,7 +508,7 @@ static void resetCardinality(NumGcInfo *info, NumericRange *range, size_t blocks
   }
   // Add the entries that were added since the fork to the HLL
-  size_t startIdx = GcScanDelta_LastBlockIdx(info->delta);
+  size_t startIdx = gc_delta_last_block_idx;
```
We can track the changing part, but it's more complicated. If we start merging or splitting blocks by the fork, we will need to track these changes as well.

The simple and robust calculation is to take the difference between the observed number of blocks before the fork and the number of blocks after (before applying the delta). Then maybe add 1 if we ignore the last block, and this is the number of trailing blocks we need to add to the HLL manually

chesedo commented:
> Can we get a regression test?

The `FGCTestNumeric.testNumericBlocksSinceFork` test already covered this code path. But it just read that freed memory without failing :sweat_smile: . I updated the code to explicitly set that pointer to `NULL` after it is freed. With this that test not fails trying to access a `NULL` pointer. This can be seen by running this test on the commit right before the fix.

```bash
./build.sh RUN_UNIT_TESTS TEST="FGCTestNumeric.testNumericBlocksSinceFork"
```

I've also updated the code to use the `NULL` state to correctly free that data.

chesedo commented on src/fork_gc.c:
```diff
@@ -498,7 +498,7 @@ static FGCError recvNumIdx(ForkGC *gc, NumGcInfo *ninfo) {
 }
 
-static void resetCardinality(NumGcInfo *info, NumericRange *range, size_t blocksSinceFork) {
+static void resetCardinality(NumGcInfo *info, NumericRange *range, size_t blocksSinceFork, size_t gc_delta_last_block_idx) {
```
I don't see a way in using only one of them.

Suppose I only use `gc_delta_last_block_idx`, then I'll have to calculate `blocksSinceFork` in `resetCardinality`. But calling `InvertedIndex_NumBlocks(range->entries)` here to make that calculation can give incorrect results since the GC apply step might have reduced the number of blocks. That's why this variable is calculated before the apply step.

Now suppose I pass `blocksSinceFork` down and remove `gc_delta_last_block_idx` which I'll then calculate in `resetCardinality`. I won't be able to call `InvertedIndex_NumBlocks(range->entries)` to get the original last block idx since the GC apply might have changed the number of inverted index blocks on this path too.

So both cases of using only one of the variables results in possible errors when calculating the other one.

GuyAv46 replied:
But you don’t need to original number of blocks, you need the trailing blocks since the fork from the current number of blocks

chesedo replied:
The other update caused me to only use one variable
